### PR TITLE
Implementing v8 conformance tests

### PR DESCRIPTION
**Pull request by czearing**
#### Pull request checklist
- [x] Related to #14494
- [X] Include a change request file using `$ yarn change`

#### Description of changes
1. Adding conformance tests to V8 components.

2. Renamed `sharedIsConformant` to `isConformant`.

4. Adding `isConformant` files to eslint configHelpers `testFiles` array.

3. Adding `isConformant` as divDependencies to existing V8 packages.

4. Adding `isConformant` file to `react-link`, `react-tabs`, `react-toggle`, `react-button`, and `react-checkbox`.

5. Adding `displayNames` to react-next components.

6. Fix to a visual regression in `ChoiceGroupOption`.

For reference here is a HackMD with a list of the conformance tests that will run by default: https://hackmd.io/4i2Xl0d8TiuZLcBAPznnnA

### V8 conformant tests added and disabledTests: 
- ### `react-next`: 'has-docblock', 'kebab-aria-attributes'
    * Breadcrumb
    * Callout
    * ChoiceGroup
    * ComboBox: `component-renders`
    * ContextualMenu
    * Dropdown
    * FocusTrapZone
    * Image
    * MessageBar
    * Modal
    * OverflowSet
    * Persona
    * PersonaCoin
    * PersonaPresence: `exported-top-level, has-top-level-file`
    * Rating
    * ResizeGroup
    * SearchBox
    * BaseSelectedItemsList: `has-top-level-file`
    * SelectedPeopleList: `has-top-level-file`
    * Shimmer
    * SpinButton
    * SwatchColorPicker
    * TeachingBubble
    * MaskedTextField: `has-top-level-file`
    * TextField
    * Grid: ` has-top-level-file`
- ### `react-button`: 'has-docblock', 'kebab-aria-attributes'
    * Button
    * MenuButton
    * SplitButton
    * ToggleButton:`
    
- ### `react-slider`: 'has-docblock', 'kebab-aria-attributes'
    * Slider
- ### `react-toggle`: 'has-docblock', 'kebab-aria-attributes'
    * Toggle
- ### `react-tabs`: 'has-docblock', 'kebab-aria-attributes'
    * Pivot
- ### `react-link`: 'has-docblock', 'kebab-aria-attributes'
    * Link
- ### `react-checkbox`: 'has-docblock', 'kebab-aria-attributes'
    * Checkbox

## V8 conformant tests that need to be addressed in the future:
- ### `react-next`: 'has-docblock', 'kebab-aria-attributes'
    * Layer
    * ContextualMenuButton
    * ContextualMenuSplitButton

### CODE REVIEW

File: packages/react-next/src/components/Callout/Callout.test.tsx [hunk 1]
```diff
@@ -47,6 +48,13 @@ describe('Callout', () => {
     });
   });
 
+  isConformant({
+    componentPath: path.join(__dirname, 'Callout.tsx'),
+    Component: Callout,
+    displayName: 'Callout',
+    disabledTests: ['component-has-displayname'],
```
Comment (ecraig12345) [inline packages/react-next/src/components/Callout/Callout.test.tsx:55]: This should definitely have a displayName or constructor name. If it doesn't today we should fix it.

Same in other component tests.

File: packages/react-toggle/src/common/isConformant.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+// eslint-disable-next-line import/no-extraneous-dependencies -- this file is for testing
```
Comment (ecraig12345) [inline packages/react-toggle/src/common/isConformant.ts:1]: Remove this

File: packages/react-next/src/components/Persona/PersonaPresence/PersonaPresence.test.tsx [hunk 3]
```diff
@@ -93,4 +95,11 @@ describe('PersonaPresence', () => {
     const tree = component.toJSON();
     expect(tree).toMatchSnapshot();
   });
+
+  isConformant({
+    componentPath: path.join(__dirname, 'PersonaPresence.tsx'),
+    Component: PersonaPresence,
+    displayName: 'PersonaPresence',
+    disabledTests: ['exported-top-level', 'has-top-level-file'],
```
Comment (ecraig12345) [inline packages/react-next/src/components/Persona/PersonaPresence/PersonaPresence.test.tsx:103]: This one actually is exported top-level (despite not having a top-level file), so the test shouldn't be disabled. If the test isn't handling that scenario properly we should fix it.

File: packages/react-next/src/components/SelectedItemsList/BaseSelectedItemsList.test.tsx [hunk 4]
```diff
@@ -28,6 +30,13 @@ describe('SelectedItemsList', () => {
       expect(tree).toMatchSnapshot();
     });
 
+    isConformant({
+      componentPath: path.join(__dirname, 'BaseSelectedItemsList.tsx'),
+      Component: BaseSelectedItemsList,
+      displayName: 'BaseSelectedItemsList',
+      disabledTests: [`exported-top-level`, `has-top-level-file`],
```
Comment (ecraig12345) [inline packages/react-next/src/components/SelectedItemsList/BaseSelectedItemsList.test.tsx:37]: This one also shouldn't have `exported-top-level` disabled

File: packages/react-next/src/components/SelectedItemsList/SelectedPeopleList/SelectedPeopleList.test.tsx [hunk 5]
```diff
@@ -1,11 +1,20 @@
...
 
 import { SelectedPeopleList, IExtendedPersonaProps } from './SelectedPeopleList';
 
 describe('SelectedPeopleList', () => {
   describe('Element keying behavior', () => {
+    isConformant({
+      componentPath: path.join(__dirname, 'SelectedPeopleList.tsx'),
+      Component: SelectedPeopleList,
+      displayName: 'SelectedPeopleList',
+      disabledTests: [`exported-top-level`, `has-top-level-file`],
```
Comment (ecraig12345) [inline packages/react-next/src/components/SelectedItemsList/SelectedPeopleList/SelectedPeopleList.test.tsx:15]: This one also shouldn't have `exported-top-level` disabled

**Review by ecraig12345 — COMMENTED** [review]
Mostly looks good! A few suggestions.

File: packages/react-button/src/common/isConformant.ts [hunk 6]
```diff
@@ -0,0 +1,11 @@
+import { isConformant as baseIsConformant, IsConformantOptions } from '@fluentui/react-conformance';
+
+export function isConformant(testInfo: Omit<IsConformantOptions, 'componentPath'>) {
+  const defaultOptions = {
+    disabledTests: ['has-docblock', 'kebab-aria-attributes', 'as-*'],
+  };
+
+  const path = { componentPath: module!.parent!.filename.replace('.test', '') };
```
Comment (ecraig12345) [inline packages/react-button/src/common/isConformant.ts:8]: It might be better to just include this in defaultOptions (same in other packages)

File: packages/react-button/src/components/SplitButton/SplitButton.test.tsx [hunk 7]
```diff
@@ -0,0 +1,10 @@
+import { SplitButton } from './SplitButton';
+import { isConformant } from '../../common/isConformant';
+
+describe('SplitButton', () => {
+  isConformant({
+    Component: SplitButton,
+    displayName: 'SplitButton',
+    disabledTests: ['has-docblock', 'as-renders-html', 'as-passes-as-value', 'as-renders-react-class', 'as-renders-fc'],
```
Comment (ecraig12345) [inline packages/react-button/src/components/SplitButton/SplitButton.test.tsx:8]: Remove these?

File: packages/react-checkbox/src/common/isConformant.ts [hunk 8]
```diff
@@ -0,0 +1,11 @@
+import { isConformant as baseIsConformant, IsConformantOptions } from '@fluentui/react-conformance';
+
+export function isConformant(testInfo: Omit<IsConformantOptions, 'componentPath'>) {
```
Comment (ecraig12345) [inline packages/react-checkbox/src/common/isConformant.ts:3]: Can you add a call to this in the checkbox tests?

File: packages/react-link/src/common/isConformant.ts [hunk 9]
```diff
@@ -0,0 +1,11 @@
+import { isConformant as baseIsConformant, IsConformantOptions } from '@fluentui/react-conformance';
+
+export function isConformant(testInfo: Omit<IsConformantOptions, 'componentPath'>) {
```
Comment (ecraig12345) [inline packages/react-link/src/common/isConformant.ts:3]: Can you call this in the link tests? (current and /next)

File: packages/react-next/src/components/ChoiceGroup/ChoiceGroup.base.tsx [hunk 10]
```diff
@@ -156,3 +145,17 @@ export const ChoiceGroupBase: React.FunctionComponent = React.forwardRef(
     );
   },
 );
+ChoiceGroupBase.displayName = COMPONENT_NAME;
+
+function useDebugWarnings(props: IChoiceGroupProps) {
+  if (process.env.NODE_ENV !== 'production') {
+    // eslint-disable-next-line react-hooks/rules-of-hooks -- build-time conditional
+    useWarnings({
+      name: 'Checkbox',
```
Comment (ecraig12345) [inline packages/react-next/src/components/ChoiceGroup/ChoiceGroup.base.tsx:154]:
```suggestion
      name: COMPONENT_NAME.replace('Base', ''),
```

File: packages/react-next/src/components/OverflowSet/OverflowSet.test.tsx [hunk 11]
```diff
@@ -69,6 +70,11 @@ describe('OverflowSet', () => {
     });
   });
 
+  isConformant({
```
Comment (ecraig12345) [inline packages/react-next/src/components/OverflowSet/OverflowSet.test.tsx:73]: Probably rather than renaming OverflowSet.ts to tsx, the test should be updated to handle either ts or tsx.

Same with other components.

File: change/@fluentui-react-button-2020-09-02-14-32-55-feat-implementing-v8-conformance-tests.json [hunk 12]
```diff
@@ -0,0 +1,8 @@
+{
+  "type": "patch",
+  "comment": "Adding conformant tests to v8 component test files.",
```
Comment (ecraig12345) [inline change/@fluentui-react-button-2020-09-02-14-32-55-feat-implementing-v8-conformance-tests.json:3]: For packages where you just changed tests, the change type should be none. (But it should be patch if you added displayNames or fixed other things.)

File: packages/react-button/src/components/SplitButton/SplitButton.test.tsx (hunk 7 above)
Comment (ecraig12345) [reply packages/react-button/src/components/SplitButton/SplitButton.test.tsx:8]: Unless I'm misunderstanding and the as tests work for the other buttons but not this one

File: packages/react-next/src/components/TextField/MaskedTextField/MaskedTextField.test.tsx [hunk 13]
```diff
@@ -13,6 +13,12 @@ describe('MaskedTextField', () => {
     expect(tree).toMatchSnapshot();
   });
 
+  isConformant({
+    Component: MaskedTextField,
+    displayName: 'MaskedTextField',
+    disabledTests: ['has-top-level-file', 'exported-top-level'],
```
Comment (ecraig12345) [inline packages/react-next/src/components/TextField/MaskedTextField/MaskedTextField.test.tsx:19]: This does appear to be exported (at least in OUFR--if it's not in react-next we should fix that)
```suggestion
    disabledTests: ['has-top-level-file'],
```

File: packages/react-checkbox/src/components/Checkbox/Checkbox.test.tsx [hunk 14]
```diff
@@ -4,6 +4,7 @@ import { create } from '@uifabric/utilities/lib/test';
 import { mount, ReactWrapper } from 'enzyme';
 
 import { Checkbox } from './Checkbox';
+import { isConformant } from '../../common/isConformant';
```
Comment (ecraig12345) [inline packages/react-checkbox/src/components/Checkbox/Checkbox.test.tsx:7]: Can you add this for the /next checkbox too?

Same for link, slider, tabs, and toggle's /next versions.

Comment (xugao) [reply packages/react-checkbox/src/components/Checkbox/Checkbox.test.tsx:7]: i decided to not duplicate the tests (yet) in `/next` folders. so no need here

File: packages/react-next/src/utilities/ButtonGrid/ButtonGrid.test.tsx [hunk 15]
```diff
@@ -43,7 +44,7 @@ describe('ButtonGrid', () => {
   });
   it('Can render a ButtonGrid with width of 2', () => {
     const wrapper = shallow(
-      <ButtonGrid
+      <ButtonGridBase
```
Comment (ecraig12345) [inline packages/react-next/src/utilities/ButtonGrid/ButtonGrid.test.tsx]: Just curious, why are these being run with ButtonGridBase? That's not what we do for most of the component tests.

File: packages/react-checkbox/src/components/Checkbox/Checkbox.test.tsx (hunk 14 above)
Comment (ecraig12345) [reply packages/react-checkbox/src/components/Checkbox/Checkbox.test.tsx:7]: Why not? To me it seems good to have the conformance tests start running early. (Though if we did that we'd probably need to add an option to react-conformance to use a different entry point file besides `src/index` for parsing.)

**Comment by ecraig12345** [comment]
@czearing There are more merge conflicts now :(

File: packages/react-checkbox/src/components/Checkbox/Checkbox.test.tsx (hunk 14 above)
Comment (xugao) [reply packages/react-checkbox/src/components/Checkbox/Checkbox.test.tsx:7]: i was trying to avoid duplication of the same code in the code base. `/next` is not something we should worry until we get v8 out of the door. we have 2 copies now, didn't want to make it 3.

File: packages/react-next/src/utilities/ButtonGrid/ButtonGrid.test.tsx (hunk 15 above)
Comment (czearing) [reply packages/react-next/src/utilities/ButtonGrid/ButtonGrid.test.tsx]: I investigated this and it is a bit of a strange scenario. It seems after switching `ButtonGridBase` to `ButtonGrid` and shallow to safeMount, the tests fail because the gridcell role on the `DefaultButton` is called 32 times on both the v7 and v8 versions of the test. What is strange is if you console.log(wrapper.html) the gridcell role only exists eight times. After switching `DefaultButton` to button, all of the tests suddenly pass. I switched both versions to use `ButtonGrid` but I’m not entirely sure of what caused the error with `DefaultButton`.

File: packages/react-slider/src/common/isConformant.ts [hunk 16]
```diff
@@ -1,8 +1,9 @@
 import { isConformant as baseIsConformant, IsConformantOptions } from '@fluentui/react-conformance';
 
-export function isConformant(testInfo: IsConformantOptions) {
+export function isConformant(testInfo: Omit<IsConformantOptions, 'componentPath'>) {
```
Comment (xugao) [inline packages/react-slider/src/common/isConformant.ts:3]: this typing change looks incorrect to me. why is this here?

Comment (xugao) [reply packages/react-slider/src/common/isConformant.ts:3]: same for other places

Comment (ecraig12345) [reply packages/react-slider/src/common/isConformant.ts:3]: Because componentPath is now being calculated automatically based on the test filename (see below)

File: packages/react-next/src/common/isConformant.ts [hunk 17]
```diff
@@ -1,8 +1,9 @@
 import { isConformant as baseIsConformant, IsConformantOptions } from '@fluentui/react-conformance';
 
-export function isConformant(testInfo: IsConformantOptions) {
+export function isConformant(testInfo: Omit<IsConformantOptions, 'componentPath'> & { componentPath?: string }) {
   const defaultOptions = {
     disabledTests: ['has-docblock', 'kebab-aria-attributes'],
```
Comment (xugao) [inline packages/react-next/src/common/isConformant.ts:5]: I am confused on why `as-*` tests are included and passing for components in `react-next`. `as` should not be supported on most components here

File: packages/react-slider/src/common/isConformant.ts (hunk 16 above)
Comment (xugao) [reply packages/react-slider/src/common/isConformant.ts:3]: then shouldn't we update `IsConformantOptions` typing instead?

File: packages/react-next/src/common/isConformant.ts (hunk 17 above)
Comment (ecraig12345) [reply packages/react-next/src/common/isConformant.ts:5]: Internally the `as-*` tests check if the component has an `as` prop, and only run if it does.

File: packages/react-slider/src/common/isConformant.ts (hunk 16 above)
Comment (ecraig12345) [reply packages/react-slider/src/common/isConformant.ts:3]: No, the `componentPath` is a project-specific detail which each project's `isConformant` wrapper may be able to provide for its own components, but the `react-conformance` package can't provide generically.

File: packages/react-button/src/components/Button/Button.test.tsx [hunk 18]
```diff
@@ -1,16 +1,16 @@
...
 import * as renderer from 'react-test-renderer';
 import { mount, ReactWrapper } from 'enzyme';
+import { isConformant } from '../../common/isConformant';
+
 import { MergeStylesProvider } from '@fluentui/react-theme-provider';
 import { ButtonProps } from './Button.types';
 
 /** Use merge-styles provider to ensure styles show up in snapshots. */
-const Button = React.forwardRef<HTMLButtonElement, ButtonProps>((props, ref) => (
+const ButtonComponent = React.forwardRef<HTMLButtonElement, ButtonProps>((props, ref) => (
```
Comment (ecraig12345) [inline packages/react-button/src/components/Button/Button.test.tsx:11]: Might be better to call this something like ButtonWrapper

File: packages/react-checkbox/src/common/isConformant.ts [hunk 19]
```diff
@@ -0,0 +1,16 @@
...
+  const defaultOptions = {
+    disabledTests: [
+      'has-docblock',
+      'kebab-aria-attributes',
+      'as-renders-fc',
```
Comment (ecraig12345) [inline packages/react-checkbox/src/common/isConformant.ts:8]: What happened with the `as` tests?

File: packages/react-next/src/components/ChoiceGroup/ChoiceGroupOption/ChoiceGroupOption.base.tsx [hunk 20]
```diff
@@ -70,7 +70,7 @@ export const ChoiceGroupOptionBase = (props: IChoiceGroupOptionProps) => {
         {imageSrc && (
           <div className={classNames.innerField}>
             <div className={classNames.imageWrapper}>
-              <Image src={selectedImageSrc} alt={imageAlt} {...imageSize} />
+              <Image src={imageSrc} alt={imageAlt} {...imageSize} />
```
Comment (ecraig12345) [inline packages/react-next/src/components/ChoiceGroup/ChoiceGroupOption/ChoiceGroupOption.base.tsx:73]: Why this change? Bug fix?

**Review by ecraig12345 — APPROVED** [review]
A couple questions but this generally looks good, thanks!

File: packages/react-link/src/common/isConformant.ts [hunk 21]
```diff
@@ -0,0 +1,17 @@
...
+  const defaultOptions = {
+    disabledTests: [
+      `has-docblock`,
+      `kebab-aria-attributes`,
+      `as-renders-fc`,
+      `as-renders-react-class`,
+      `as-passes-as-value`,
```
Comment (xugao) [inline packages/react-link/src/common/isConformant.ts:10]: i dont think we should have these disabled.

Comment (czearing) [reply packages/react-link/src/common/isConformant.ts:10]: There seems to be a function component in link that is not using forwardRef. It could be possible that the old compose link is using is responsible, but I will dig in more. Here is the error isConformant runs into after link runs the as tests:
```  jsx
    [console.error] Error: Uncaught [Error: [console.error] Warning: Function components cannot be given refs. Attempts to access this ref will fail. Did you mean to use React.forwardRef()?%s%s

    Check the render method of `LinkBase`.
        in MyComponent (created by LinkBase)
        in LinkBase (created by StyledLinkBase)
        in StyledLinkBase (created by WrapperComponent)
        in WrapperComponent]

      19 |     throw args[0];
      20 |   } else {
    > 21 |     throw new Error(`[console.${type}] ${args.join(' ')}`);
         |           ^
      22 |   }
      23 | }
      24 |
```

File: packages/react-next/src/components/ChoiceGroup/ChoiceGroupOption/ChoiceGroupOption.base.tsx (hunk 20 above)
Comment (czearing) [reply packages/react-next/src/components/ChoiceGroup/ChoiceGroupOption/ChoiceGroupOption.base.tsx:73]: It is an intended bug fix that screener noticed in this pr. `ChoiceGroupOption`’s is supposed to have an Image src for both imageSrc and selectedImageSrc props to allow for a unique selection image. Here's what the fix does:

Without Fix:
image

Fix:
image

File: packages/react-next/src/components/ComboBox/ComboBox.test.tsx [hunk 22]
```diff
@@ -68,6 +69,18 @@ describe('ComboBox', () => {
...
+  it(`renders`, () => {
+    safeCreate(<ComboBox options={DEFAULT_OPTIONS} />, wrapper => {
+      expect(wrapper.root).toBeDefined();
+    });
+  });
+
+  isConformant({
+    Component: ComboBox,
+    displayName: 'ComboBox',
+    disabledTests: ['component-renders'],
```
Comment (ecraig12345) [inline packages/react-next/src/components/ComboBox/ComboBox.test.tsx:81]: Can you add a comment about why this had to be disabled?

File: packages/react-link/src/components/Link/Link.test.tsx [hunk 23]
```diff
@@ -66,6 +69,12 @@ describe('Link', () => {
     expect(tree).toMatchSnapshot();
   });
 
+  isConformant({
+    Component: Link,
+    displayName: 'Link',
+    disabledTests: [`as-renders-fc`, `as-passes-as-value`],
```
Comment (ecraig12345) [inline packages/react-link/src/components/Link/Link.test.tsx:75]: Would also be good to comment about why these are disabled.

**Review by ecraig12345 — APPROVED** [review]
Approved once the comments on disabled tests have been added. Thanks!

File: packages/react-link/src/common/isConformant.ts (hunk 21 above)
Comment (czearing) [reply packages/react-link/src/common/isConformant.ts]: I figured out what was causing the remaining errors and the 'as' tests are now enabled in `link`.

**Comment by ecraig12345** [comment]
Looks like you're missing a change file, and I'm seeing some weird failure from isConformant in react-northstar.

File: packages/react-toggle/src/components/Toggle/Toggle.test.tsx [hunk 24]
```diff
@@ -20,6 +21,12 @@ describe('Toggle', () => {
     ).toEqual('Label');
   });
 
+  isConformant({
+    Component: Toggle,
+    displayName: 'Toggle',
+    disabledTests: ['as-passes-as-value'],
```
Comment (xugao) [inline packages/react-toggle/src/components/Toggle/Toggle.test.tsx:27]: we should not need this one?

**Review by xugao — APPROVED** [review]
Thank you  for making this change!
